promql/prettier/node.go

```diff
@@ -0,0 +1,163 @@
+package prettier
+
+import (
+	"github.com/prometheus/prometheus/promql/parser"
+	"reflect"
+)
+
+type nodeInfo struct {
+	head parser.Expr
+	// Node details.
+	columnLimit int
+	history     []reflect.Type
+	item        parser.Item
+}
+
+func newNodeInfo(head parser.Expr, columnLimit int, item parser.Item) *nodeInfo {
+	return &nodeInfo{
+		head:        head,
+		columnLimit: columnLimit,
+		item:        item,
+	}
+}
+
+func (p *nodeInfo) violatesColumnLimit() bool {
+	return len(p.head.String()) > p.columnLimit
+}
+
+// getNode returns the node corresponding to the given position range in the AST.
+func (p *nodeInfo) getNode(root parser.Expr, item parser.Item) parser.Expr {
+	history, node, _ := p.nodeHistory(root, item.PositionRange(), []reflect.Type{})
+	p.history = history
+	return node
+}
+
+func (p *nodeInfo) baseIndent(item parser.Item) int {
+	history, _, _ := p.nodeHistory(p.head, item.PositionRange(), []reflect.Type{})
+	history = reduceContinuous(history, "*parser.BinaryExpr")
+	return len(history)
+}
+
+// contains verifies whether the current node contains a particular entity.
+func (p *nodeInfo) contains(element string) bool {
+	switch element {
+	case "grouping-modifier":
+		switch n := p.head.(type) {
+		case *parser.BinaryExpr:
+			return len(n.VectorMatching.MatchingLabels) > 0 || n.ReturnBool
+		}
+	case "scalars":
+		switch n := p.head.(type) {
+		case *parser.BinaryExpr:
+			if n.LHS.Type() == parser.ValueTypeScalar || n.RHS.Type() == parser.ValueTypeScalar {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// parentNode returns the parent node of the given node/item position range.
+func (p *nodeInfo) parentNode(head parser.Expr, rnge parser.PositionRange) reflect.Type {
+	ancestors, _, found := p.nodeHistory(head, rnge, []reflect.Type{})
+	if !found {
+		return nil
+	}
+	if len(ancestors) < 2 {
+		return nil
+	}
+	return ancestors[len(ancestors)-2]
+}
+
+// nodeHistory returns the ancestors of the node the item position range is passed of,
+// along with the node in which the item is present. This is done with the help of AST.
+// posRange can also be called as itemPosRange since carries the position range of the lexical item.
+func (p *nodeInfo) nodeHistory(head parser.Expr, posRange parser.PositionRange, stack []reflect.Type) ([]reflect.Type, parser.Expr, bool) {
```

Comment:
Would it make sense to work with slices of `parser.Node` instead of `reflect.Type` here?

Reply:
How comfortable will it be for comparing types of 2 nodes in `parser.Node`?

Reply:
When would you need to do such a thing?

Reply:
Addressed in recent commits. Hence, resolving.